File: algorithms_and_data_structure/BinarySearch2/main.go
package main

import "fmt"

func helper(array []int, target, left, right int) int {
	if left > right {
		return -1
	}

	middle := (left + right) / 2
	potentialMatch := array[middle]

	if target == potentialMatch {
		return middle
	} else if target < potentialMatch {
		return helper(array, target, left, middle-1)
	}
	return helper(array, target, middle+1, right)
}

func main() {
	array := []int{1, 3, 5, 7, 9, 11, 13, 15, 17}
	targets := []int{7, 15, 2, 11}

	for _, target := range targets {
		index := helper(array, target, 0, len(array)-1)
		if index != -1 {
			fmt.Printf("Target %d found at index %d\n", target, index)
		} else {
			fmt.Printf("Target %d not found\n", target)
		}
	}
}

File: algorithms_and_data_structure/BinarySearch2/main.py
def binary_search_recursive(array, target):
    def helper(left, right):
        if left > right:
            return -1
        mid = (left + right) // 2
        if array[mid] == target:
            return mid
        elif array[mid] > target:
            return helper(left, mid - 1)
        else:
            return helper(mid + 1, right)
    return helper(0, len(array) - 1)

